node/pkg/cni/token_watch.go

```diff
@@ -1,56 +1,245 @@
 package cni
 
 import (
+	"context"
 	"encoding/base64"
+	"encoding/json"
 	"fmt"
 	"io/ioutil"
+	"math/rand"
 	"os"
+	"strings"
+	"sync"
 	"time"
 
 	"github.com/sirupsen/logrus"
-	"gopkg.in/fsnotify/fsnotify.v1"
+	authenticationv1 "k8s.io/api/authentication/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/rest"
 )
 
-var serviceaccountDirectory string = "/var/run/secrets/kubernetes.io/serviceaccount/"
+const DefaultServiceAccountName = "calico-node"
+const serviceAccountNamespace = "/var/run/secrets/kubernetes.io/serviceaccount/namespace"
+const tokenFile = "/var/run/secrets/kubernetes.io/serviceaccount/token"
+const defaultCNITokenValiditySeconds = 24 * 60 * 60
+const minTokenRetryDuration = 5 * time.Second
+const defaultRefreshFraction = 4
+
 var kubeconfigPath string = "/host/etc/cni/net.d/calico-kubeconfig"
 
-func Run() {
-	// Log to stdout.  this prevents our logs from being interpreted as errors by, for example,
-	// fluentd's default configuration.
-	logrus.SetOutput(os.Stdout)
+type TokenRefresher struct {
+	tokenSupported bool
+	tokenOnce      *sync.Once
+
+	tokenValiditySeconds   int64
+	minTokenRetryDuration  time.Duration
+	defaultRefreshFraction time.Duration
+
+	clientset *kubernetes.Clientset
+
+	namespace          string
+	serviceAccountName string
+
+	tokenChan chan TokenUpdate
+}
+
+type TokenUpdate struct {
+	Token          string
+	ExpirationTime time.Time
+}
 
-	// Create a watcher for file changes.
-	watcher, err := fsnotify.NewWatcher()
+func NamespaceOfUsedServiceAccount() string {
+	namespace, err := ioutil.ReadFile(serviceAccountNamespace)
 	if err != nil {
-		panic(err)
+		logrus.WithError(err).Fatal("Failed to read service account namespace file")
 	}
-	defer watcher.Close()
+	return string(namespace)
+}
 
-	// Watch for changes to the serviceaccount directory. Rety if necessary.
-	for {
-		if err := watcher.Add(serviceaccountDirectory); err != nil {
-			// Error watching the file - retry
-			logrus.WithError(err).Error("Failed to watch Kubernetes serviceaccount files.")
-			time.Sleep(5 * time.Second)
-			continue
+func InClusterClientSet() (*kubernetes.Clientset, error) {
```

Comment:
Given this also supports an explicit KUBECONFIG env var, should probably drop the "InCluster" from the name, and just call this "BuildClientSet" or something like that.

Reply:
Done.